### PR TITLE
Add Task and TaskExecutor @open sesame 11/11 10:50

### PR DESCRIPTION
**Pull request by jihochu**
Task has several properties for its state and working information.
There are two types for Task, Synchronous and Asynchronous. and, their
behaviors will be differ when they are executed by TaskExecutor.

TaskExecutor class manages task's execution by their type.
Synchoronous tasks are exeucuted immediately, while asynchoronous tasks
are executed in thread manner.

For testing above two classes, unittest codes are also included.
The mock class for TaskExecutor is introduced to check work thread is
correctly working, and completion check is also working correctly.
Many tests are added to check asynchronous tasks' correctness.

Issue (#2034)

Signed-off-by: Jiho Chu <jiho.chu@samsung.com>

### CODE REVIEW

File: nntrainer/tensor/task_executor.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+  /**
+   * @brief Check task is completed
+   *
+   * @param data user data
+   */
+  template <typename T>
+  void completeChecker(int id, std::shared_ptr<TaskAsync<T>> task) {
+    // wait until tasks are fully mapped
+    while (tasks.find(id) == tasks.end())
```
Comment (jijoongmoon) [inline nntrainer/tensor/task_executor.h:135]: I wonder if there is possiblity to loop forever when the id is greater than  ```map::size()```. better to check if the id is less than ```map::size()```

Comment (jihochu) [reply nntrainer/tensor/task_executor.h:135]: Thanks.
timeout logic is added to. max timeout is set as 1 sec.

File: nntrainer/tensor/task_executor.h [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+class TaskExecutor {
+public:
+  enum CompleteStatus {
+    SUCCESS,
+    FAIL_CANCEL,
+    FAIL_TIMEOUT,
+    FAIL,
+  };
+
+  using CompleteCallback =
```
Comment (jijoongmoon) [inline nntrainer/tensor/task_executor.h:42]: Could you add more explanation as an comment?

File: nntrainer/tensor/task_executor.h [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    NNTR_THROW_IF(tasks.find(id) == tasks.end(), std::runtime_error)
+      << "tasks was not correctly mapped";
+
+    auto status = std::get<0>(tasks[id]).wait_for(T(task->getTimeout()));
+    handleCompleteStatus(id, status);
+
+    task->setState(Task::State::DONE);
+  }
+
```
Comment (jijoongmoon) [inline nntrainer/tensor/task_executor.h:150]: add doxygen comment for this function?

**Review by jijoongmoon — APPROVED** [review]
LGTM except minor comment.

File: nntrainer/tensor/task_executor.h (hunk 2 above)
Comment (jihochu) [reply nntrainer/tensor/task_executor.h:48]: thanks. it added.

File: nntrainer/tensor/task_executor.h (hunk 3 above)
Comment (jihochu) [reply nntrainer/tensor/task_executor.h:157]: doxy comments are added.

**Review by DonghakPark — APPROVED** [review]
LGTM!

**Review by songgot — APPROVED** [review]
LGTM

**Pull request merged**